moveit_planners/trajopt/src/trajopt_interface.cpp

```diff
@@ -0,0 +1,458 @@
+  ros::WallTime start_time = ros::WallTime::now();
+  robot_model::RobotModelConstPtr robot_model = planning_scene->getRobotModel();
+  robot_state::RobotStatePtr current_state(new robot_state::RobotState(robot_model));
+  *current_state = planning_scene->getCurrentState();
+  const robot_state::JointModelGroup* joint_model_group = current_state->getJointModelGroup(req.group_name);
+  std::vector<std::string> joint_names = joint_model_group->getVariableNames();
+  int dof = joint_names.size();
+  std::vector<double> current_joint_values;
+  current_state->copyJointGroupPositions(joint_model_group, current_joint_values);
+
+  // current state is different from star state in general
+  ROS_INFO(" ======================================= Extract start stae infromation");
+  std::unordered_map<std::string, double> all_joints;
+  trajopt::DblVec joint_start_constraints;
+
+  for (int joint_index = 0; joint_index < req.start_state.joint_state.position.size(); ++joint_index)
+  {
+    all_joints[req.start_state.joint_state.name[joint_index]] = req.start_state.joint_state.position[joint_index];
+  }
+
+  for (auto joint_name : joint_names)
+  {
+    ROS_INFO(" joint position from start state, name: %s, value: %f", joint_name.c_str(), all_joints[joint_name]);
+    joint_start_constraints.push_back(all_joints[joint_name]);
+  }
+
+  // check the joint limit violiation for star state
+  if (not planning_scene->getRobotModel()->satisfiesPositionBounds(joint_start_constraints.data()))
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "Start state violates joint limits");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_ROBOT_STATE;
+    return false;
+  }
+
+  ROS_INFO(" ======================================= Create ProblemInfo");
+  ProblemInfo problem_info(planning_scene, req.group_name);
+
+  setProblemInfoParam(problem_info);
+
+  ROS_INFO(" ======================================= Populate init info, hard-coded");
+  // TODO: init info should be defined by user. To this end, we need to add seed trajectories to MotionPlanRequest.
+  // JOINT_INTERPOLATED: data is the current joint values
+  // GIVEN_TRAJ: data is the joint values of the current state copied to all timesteps
+  Eigen::VectorXd current_joint_values_eigen(dof);
+  Eigen::VectorXd start_joint_values_eigen(dof);
+  for (int joint_index = 0; joint_index < dof; ++joint_index)
+  {
+    current_joint_values_eigen(joint_index) = current_joint_values[joint_index];
+  }
+
+  if (problem_info.init_info.type == InitInfo::JOINT_INTERPOLATED)
+  {
+    problem_info.init_info.data = current_joint_values_eigen;
+  }
+  else if (problem_info.init_info.type == InitInfo::GIVEN_TRAJ)
+  {
+    problem_info.init_info.data = current_joint_values_eigen.transpose().replicate(problem_info.basic_info.n_steps, 1);
+  }
+
+  ROS_INFO(" ======================================= Create Constraints");
+  if (req.goal_constraints.empty())
```

Comment:
This would be more readable if you would move some code into separate functions. For instance this part could just go into a function`extractGoalConstraints()`.

Reply:
added extractStartJointValues(req)